### PR TITLE
Return ResultAsync<T, E> from safeTry instead of Promise<Result<T, E>> for better composability

### PR DESCRIPTION
**Pull request by sharno**

### CODE REVIEW

File: .changeset/safeTryAsyncResult.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'neverthrow': patch
```
Comment (mattpocock) [inline .changeset/safeTryAsyncResult.md:2]: Would likely need to be a minor?

File: src/result.ts [hunk 2]
```diff
@@ -104,26 +104,24 @@ export function safeTry<
...
-): Promise<
-  Result<
-    InferOkTypes<GeneratorReturnResult>,
-    InferErrTypes<YieldErr> | InferErrTypes<GeneratorReturnResult>
-  >
+): ResultAsync<
+  InferOkTypes<GeneratorReturnResult>,
+  InferErrTypes<YieldErr> | InferErrTypes<GeneratorReturnResult>
 >
 export function safeTry<T, E>(
```
Comment (mattpocock) [inline src/result.ts:117]: Could this become a set of function overloads?

I.e. one branch takes a generator, returns a Result. The other takes an async generator, returns ResultAsync.

**Comment by mattpocock** [comment]
Appears to be a fix for #561

**Comment by sharno** [comment]
@mattpocock 
It missed this comment written right above the code that this PR is changing
```
// NOTE:
// Since body is potentially throwable because `await` can be used in it,
// Promise<Result<T, E>>, not ResultAsync<T, E>, is used as the return type.
```

not sure if I understand it fully

File: src/result.ts (hunk 2 above)
Comment (sharno) [reply src/result.ts:117]: It's already like this

```ts
export function safeTry<T, E>(body: () => Generator<Err<never, E>, Result<T, E>>): Result<T, E>
```
and
```ts
export function safeTry<T, E>(
  body: () => AsyncGenerator<Err<never, E>, Result<T, E>>,
): ResultAsync<T, E>
```

**Comment by mattpocock** [comment]
@sharno Agree. Neverthrow doesn't guarantee that a function will never throw. So this feels overly defensive on the library's part.

**Review by supermacro — APPROVED** [review]

**Comment by supermacro** [comment]
@sharno looks good, but note that types need to be updated on some tests

**Comment by sharno** [comment]
updated and fixed the tests 👍

**Review by supermacro — APPROVED** [review]

**Pull request merged**